UnitTestFrameworkPkg/Library/Posix/MemoryAllocationLibPosix/AllocateBelowAddress.c

```diff
@@ -0,0 +1,435 @@
+/** @file
+  Instance of Memory Below Address Allocation Library based on Windows APIs
+  and Linux APIs.
+
+  Uses Windows APIs VirtualAlloc() and VirtualFree() to allocate and free memory
+  below a specified virtual address.
+
+  Uses Linux APIs mmap() and munmap() to allocate and free memory below a
+  specified virtual address.
+
+  Copyright (c) 2024, Intel Corporation. All rights reserved.<BR>
+  SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#if defined (_WIN32) || defined (_WIN64)
+  #include "WinInclude.h"
+#elif defined (__linux__)
+  #include <sys/mman.h>
+  #include <unistd.h>
+  #include <errno.h>
+  #include <string.h>
+#else
+  #error Unsupported target
+#endif
+
+#include <Library/HostMemoryAllocationBelowAddressLib.h>
+#include <Library/DebugLib.h>
+
+///
+/// Signature for PAGE_HEAD_BELOW_ADDRESS structure
+/// Used to verify that buffer being freed was allocated by this library.
+///
+#define PAGE_HEAD_BELOW_ADDRESS_PRIVATE_SIGNATURE  SIGNATURE_64 ('P', 'A', 'H', 'B', 'e', 'l', 'A', 'd')
+
+///
+/// Structure placed immediately before an aligned allocation to store the
+/// information required to free the entire allocated buffer.
+///
+typedef struct {
+  UINT64    Signature;
+  VOID      *AllocatedBuffer;
+  UINTN     TotalPages;
+  VOID      *AlignedBuffer;
+  UINTN     AlignedPages;
+} PAGE_HEAD_BELOW_ADDRESS;
+
+///
+/// Signature for POOL_HEAD_BELOW_ADDRESS structure
+/// Used to verify that buffer being freed was allocated by this library.
+///
+#define POOL_HEAD_BELOW_ADDRESS_PRIVATE_SIGNATURE  SIGNATURE_64 ('P', 'O', 'H', 'B', 'e', 'l', 'A', 'd')
+
+///
+/// Structure placed immediately before an pool allocation to store the
+/// information required to free the entire allocated buffer.
+///
+typedef struct {
+  UINT64    Signature;
+  UINT64    TotalSize;
+} POOL_HEAD_BELOW_ADDRESS;
+
+//
+// Lowest address that can be allocated by this library
+//
+#define MINIMUM_ALLOCATION_ADDRESS  BASE_64KB
+
+//
+// The page size of the host
+//
+static UINTN  mPageSize = 0;
+
+/**
+  Use system services to get the host page size.
+
+  @return  Host page size in bytes.
+**/
+static

```

Comment:
Some new functions added are using `static` (like this instance) while others use `STATIC` like this instance in `UnitTestLib.c`. Is that intentional?

```c
STATIC
EFI_STATUS
FreeUnitTestTestEntry (
  IN UNIT_TEST_LIST_ENTRY  *TestEntry
  )
```

Reply:
Yes.  The recommended code style is to use `static`.  I am doing this in new code.  Updating existing source will take a while.